### PR TITLE
Connect to API Gateway 

### PR DESCRIPTION
**Pull request by kevmo**
This PR represents our work connecting with the API gateway in a secure way.  Some code/comments are there to mark places that need work once development from the API team is finished (e.g. the unique number header we are expecting from UIO, an API response with end user information in it rather than a placeholder).  

---

Environmental variables (in a root-level `.env` file) are used to make the code more configurable in the event of URL/header/parameter changes in the future. 

===

Resolves #XXX

- [x] Documentation (e.g. READMEs) updated, if relevant

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,7 +16,11 @@
   },
   "dependencies": {
     "bootstrap": "^4.6.0",
+    "dotenv": "^9.0.2",
     "next": "10.0.8",
+    "node-fetch": "^2.6.1",
+    "node-forge": "^0.10.0",
```
Comment (kevmo) [inline package.json:22]: should be removed.

File: components/ClaimCard.tsx [hunk 2]
```diff
@@ -14,6 +19,7 @@ export const ClaimCard: React.FC = () => {
           'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.',
         ]}
       />
+      <h1>{uniqueNumber}</h1>
```
Comment (kalvinwang) [inline components/ClaimCard.tsx:22]: instead of passing the uniqueNumber to a component, let's use the logger library @lomky just added to log it on the backend whenever we make a request to the API gateway! we don't otherwise need the UN on the front-end so better not to display it.

i'd suggest in terms of just getting this working on test, just strip it out for now, and use the logger library in the next PR

File: pages/_app.tsx [hunk 3]
```diff
@@ -3,6 +3,12 @@ import { AppProps } from 'next/app'
 import { ReactElement } from 'react'
 import { appWithTranslation } from 'next-i18next'
 
+import { config } from 'dotenv'
+import { resolve } from 'path'
+
+// TODO: Make this path resolution more robust (no relative paths!)
+config({path: resolve(process.cwd(), '..','.env')})
```
Comment (kalvinwang) [inline pages/_app.tsx:10]: mind adding a comment about why this is necessary? i assume it was working before and this makes it work for some additional case?...

File: pages/claimstatus/index.tsx [hunk 4]
```diff
@@ -1,15 +1,32 @@
...
+import fetch from 'node-fetch'
 
 import { Header } from '../../components/Header'
 import { Main } from '../../components/Main'
 import { Footer } from '../../components/Footer'
 
-export default function Home(): ReactElement {
+
+export interface Claim {
+  ClaimType: (string | "not working")
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:21]: ClaimType is referenced here and in another diff line, I can't tell why since it's not used. Can remove?

File: pages/claimstatus/index.tsx [hunk 5]
```diff
@@ -1,15 +1,32 @@
...
+}
+
+export interface HomeProps {
+  uniqueNumber?: string
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:25]: let's remove this, see earlier comment on uniqueNumber

File: pages/claimstatus/index.tsx [hunk 6]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
         <title>{t('title')}</title>
         <link rel="icon" href="/favicon.ico" />
       </Head>
-
       <Header />
       <Main />
       <Footer />
+      {console.log("-Test Response-")} {/* testing, to be removed before launch */}
+      {console.dir({claimData})} {/* testing */}
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:42]: huh cool i've never used console.dir

File: pages/claimstatus/index.tsx [hunk 7]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
         <title>{t('title')}</title>
         <link rel="icon" href="/favicon.ico" />
       </Head>
-
       <Header />
       <Main />
       <Footer />
+      {console.log("-Test Response-")} {/* testing, to be removed before launch */}
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:41]: can you add a TODO to this line so it's easy to search for TODOs? thanks!

File: pages/claimstatus/index.tsx [hunk 8]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
...
-export const getServerSideProps: GetServerSideProps = async ({ locale }) => ({
-  props: {
-    ...(await serverSideTranslations(locale || 'en', ['common', 'header', 'footer'])),
-  },
-})
+export const getServerSideProps: GetServerSideProps = async ({ req, locale }) => {
+  const PASSWORD = process.env.CERTIFICATE_PASSPHRASE
+  const API_URL: string = process.env.API_URL || "fail gracefully" //TODO: Fail gracefully. 
+  // const UNIQUE_NUMBER_HEADER: string = process.env.UNIQUE_NUMBER_HEADER_TITLE || "x-unique-number"
+  const CERT_DIR: (string) = (process.env.CERTIFICATE_DIR || "/var/ssl")
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:51]: should there be a `|| "var/ssl"` here?

File: pages/claimstatus/index.tsx [hunk 9]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
...
 }
 
-export const getServerSideProps: GetServerSideProps = async ({ locale }) => ({
-  props: {
-    ...(await serverSideTranslations(locale || 'en', ['common', 'header', 'footer'])),
-  },
-})
+export const getServerSideProps: GetServerSideProps = async ({ req, locale }) => {
+  const PASSWORD = process.env.CERTIFICATE_PASSPHRASE
+  const API_URL: string = process.env.API_URL || "fail gracefully" //TODO: Fail gracefully.
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:49]: I think you can create a separate ticket for "implement error and alert when any critical environment variable does not exist", point to these lines, and move on for this PR. if the env vars for the API Gateway don't exist, that's an urgent issue.

File: pages/claimstatus/index.tsx [hunk 10]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
...
-  props: {
-    ...(await serverSideTranslations(locale || 'en', ['common', 'header', 'footer'])),
-  },
-})
+export const getServerSideProps: GetServerSideProps = async ({ req, locale }) => {
+  const PASSWORD = process.env.CERTIFICATE_PASSPHRASE
+  const API_URL: string = process.env.API_URL || "fail gracefully" //TODO: Fail gracefully. 
+  // const UNIQUE_NUMBER_HEADER: string = process.env.UNIQUE_NUMBER_HEADER_TITLE || "x-unique-number"
+  const CERT_DIR: (string) = (process.env.CERTIFICATE_DIR || "/var/ssl")
+  const KEY_PATH: string = path.join(CERT_DIR, '/ct.p12')
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:52]: i'd lean toward putting the filename in the .env?

File: pages/claimstatus/index.tsx [hunk 11]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
...
+    const options = {
+      cert: certificate.cert,
+      key: certificate.key,
+      rejectUnauthorized: false, 
+      keepAlive: false
+    };
+    const sslConfiguredAgent = new https.Agent(options);
+  
+    try {
+      const response = await fetch(API_URL, {
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:95]: i think this should be something like:
```
fetch(API_URL + new URLSearchParams({
    uniqueNumber: [get this from the incoming request header named idmUniqueNumber],
    user_key: USER_KEY,
}))
```

File: pages/claimstatus/index.tsx [hunk 12]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
...
+  async function getCertificate() {
+    const pemReadPkcs12 = promisify(pem.readPkcs12)
+    const pfx = fs.readFileSync(KEY_PATH);
+
+    const cert = await pemReadPkcs12(pfx, { p12Password: PASSWORD })
+
+    return cert;
+  }
+  
+  // TODO: Acquire uniqueNumber from header and use it to build API string
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:66]: this should be part of this PR if possible (see later comment about fetch)

File: pages/claimstatus/index.tsx [hunk 13]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
...
+    const pfx = fs.readFileSync(KEY_PATH);
+
+    const cert = await pemReadPkcs12(pfx, { p12Password: PASSWORD })
+
+    return cert;
+  }
+  
+  // TODO: Acquire uniqueNumber from header and use it to build API string 
+  // that will eventually be passed to makeRequest. 
+  // TODO: if no uniqueNumber, redirect.
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:68]: doesn't need to block this PR

File: pages/claimstatus/index.tsx [hunk 14]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
...
+  // function buildApiUrl(uniqueNumber, USER_KEY){}
+
+  async function makeRequest(certificate) {
+    const headers = {
+      Accept: 'application/json'
+    };
+
+    /* https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options
+    
+    // TODO: Store certs in memory, instead of loading every time.
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:80]: create a ticket for this, with context?

File: pages/claimstatus/index.tsx [hunk 15]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
...
+
+  async function makeRequest(certificate) {
+    const headers = {
+      Accept: 'application/json'
+    };
+
+    /* https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options
+    
+    // TODO: Store certs in memory, instead of loading every time. 
+    // TODO: rejectUnauthorized - set to true or implement a `checkServerIdentity`
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:81]: create a ticket for this? also would like @rocketnova to review this tomorrow

File: pages/claimstatus/index.tsx [hunk 16]
```diff
@@ -18,16 +35,87 @@ export default function Home(): ReactElement {
...
+  }
+
+  // where it comes together
+  const certificate = await getCertificate()
+  const data = await makeRequest(certificate)
+
+  return ({
+    props: {
+      claimData: [data],
+      ...(await serverSideTranslations(locale || 'en', ['common', 'header', 'footer'])),
```
Comment (kalvinwang) [inline pages/claimstatus/index.tsx:118]: just curious why was this necessary to add / is this something we do elsewhere?

**Review by kalvinwang — COMMENTED** [review]
Thanks @kevmo! Added some comments and questions.

File: components/ClaimCard.tsx (hunk 2 above)
Comment (kevmo) [reply components/ClaimCard.tsx:22]: Oo yeah, good catch.

File: pages/claimstatus/index.tsx (hunk 8 above)
Comment (kevmo) [reply pages/claimstatus/index.tsx:51]: It's not the worst thing in the world, but I agree it should probably be removed.

File: .gitignore [hunk 17]
```diff
@@ -1,5 +1,8 @@
 # See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
 
+# secrets
+/certs
```
[the remark this answers is not in the document]
Comment (kevmo) [reply .gitignore:4]: I am actually going to delete this, but I added some notes to 1PW and more extensive ones in a GDoc I will share.

File: pages/_app.tsx (hunk 3 above)
Comment (kevmo) [reply pages/_app.tsx:10]: Although we had an existing `.env` filed, it was solely being used for Storybook purposes, and that software looks for a .env file.  In our case, we want to use the environmental variables in app code, so dotenv is being used to expose them on the node's global `process` variable (the `process.env` object, to be specific).  

https://www.chromatic.com/docs/cli

Comment updated!

File: components/ClaimCard.tsx [hunk 18]
```diff
@@ -1,7 +1,12 @@
 import { ClaimCardDetails } from './ClaimCardDetails'
 import { ClaimCardBody } from './ClaimCardBody'
 
-export const ClaimCard: React.FC = () => {
+export interface ClaimCardProps {
+  uniqueNumber?: string
```
[the remark this answers is not in the document]
Comment (kevmo) [reply components/ClaimCard.tsx:5]: Great question. I am not sure, but it worked (I am removing uniqueNumber here).

File: pages/claimstatus/index.tsx (hunk 4 above)
Comment (kevmo) [reply pages/claimstatus/index.tsx:21]: ClaimType is a field in the API responses we get back that are passed to the React component.  This is TypeScript at work, basically.

File: pages/claimstatus/index.tsx (hunk 9 above)
Comment (kevmo) [reply pages/claimstatus/index.tsx:48]: https://github.com/cagov/ui-claim-tracker/issues/176

File: pages/claimstatus/index.tsx (hunk 15 above)
Comment (kevmo) [reply pages/claimstatus/index.tsx:91]: https://github.com/cagov/ui-claim-tracker/issues/182

File: pages/claimstatus/index.tsx (hunk 16 above)
Comment (lomky) [reply pages/claimstatus/index.tsx:136]: the new content here is the `claimData: [data]` - everything else is the existing app!

File: README.md [hunk 19]
```diff
@@ -6,19 +6,30 @@ relates to weekly certification
 ## Running the Application
 
 **Prerequisites:**
-  - Node 12+
-  - yarn
+
```
Comment (rocketnova) [inline README.md:9]: Please provide feedback if you have thoughts about how to structure this info in the README. Very open to alternate documentation suggestions.

File: pages/index.tsx [hunk 20]
```diff
@@ -19,21 +34,136 @@ export default function Home(): ReactElement {
...
+}
+
 export const getServerSideProps: GetServerSideProps = async ({ req, locale }) => {
   const isProd = process.env.NODE_ENV === 'production'
   const logger = isProd ? pino({}) : pino({ prettyPrint: true })
   logger.info(req)
 
+  /*
+   * Load environment variables to be used for authentication & API calls.
+   * @TODO: Handle error case where env vars are null or undefined.
```
Comment (rocketnova) [inline pages/index.tsx:67]: I removed the previous non-null assertions because it was failing the eslint [no-non-null-assertion](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-non-null-assertion.md) rule, which is [recommended](https://github.com/typescript-eslint/typescript-eslint/tree/master/packages/eslint-plugin). Here's a [writeup](https://medium.com/@tar.viturawong/a-note-on-typescript-non-null-assertion-operator-bad8fbe62bba) on what the non-null assertions were doing, whiche was essentially allowing typescript to ignore the fact that these might actually be undefined. Instead of ignoring, we should actively handle that error case (in a future ticket).

For now, I'm doing some [nullish coalescing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator) and just setting to empty string if the env var is undefined/null.

**Review by rocketnova — COMMENTED** [review]
@kalvinwang I think this is ready for review

File: README.md (hunk 19 above)
Comment (rocketnova) [reply README.md:9]: CC @lomky

Comment (kalvinwang) [reply README.md:9]: I'm confused-- are you removing these because they aren't prerequisites?...

File: pages/index.tsx (hunk 20 above)
Comment (kalvinwang) [reply pages/index.tsx:67]: Great, thank you!

File: README.md (hunk 19 above)
Comment (kalvinwang) [reply README.md:9]: Oh I see, the github preview made it look like these lines were being removed, but they're actually just being de-dented.

Comment (kalvinwang) [reply README.md:9]: Added docs look good, thanks!

Comment (lomky) [reply README.md:9]: Looks good to me - I think the README is still short enough that the flat structure works but I'm open to more structure if it gets out of hand

File: .eslintrc.js [hunk 21]
```diff
@@ -42,5 +42,6 @@ module.exports = {
   rules: {
     "react/react-in-jsx-scope": "off", //React is provided by NextJS instead
     "react/prop-types": "off", //strict TypeScript provides the same benefits
+    "camelcase": ["error", {"allow": ["user_key"]}], // necessary API parameter - index.tsx
```
Comment (kalvinwang) [inline .eslintrc.js:45]:
```suggestion
    "camelcase": ["error", {"allow": ["user_key"]}], // parameter named by external API gateway and used in index.tsx
```

File: README.md [hunk 22]
```diff
@@ -6,19 +6,30 @@ relates to weekly certification
...
+4. Run `yarn dev`
+5. Open [http://localhost:3000/claimstatus](http://localhost:3000/claimstatus) with your browser to see the result
 
-```bash
-yarn install
-yarn dev
-```
+### Environment Variables
+
+- ID_HEADER_NAME: The name of the header that contains the unique ID in the incoming request
```
Comment (kalvinwang) [inline README.md:23]:
```suggestion
- ID_HEADER_NAME: The name of the header that contains the EDD-defined unique ID / "Unique Number" in the incoming request
```

File: pages/index.tsx [hunk 23]
```diff
@@ -19,21 +34,136 @@ export default function Home(): ReactElement {
...
+  async function getCertificate() {
+    const pemReadPkcs12 = promisify(pem.readPkcs12)
+    const pfx = fs.readFileSync(P12_PATH)
+
+    // TS does not play very nicely with util.promisify
+    // See, e.g., https://github.com/Microsoft/TypeScript/issues/26048
+    // Non-MVP TODO: Consider removing this ignore & TypeScriptifying.
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore -- TypeScript does not handle promisify well.
+    const keybundle = await pemReadPkcs12(pfx, {})
```
Comment (kalvinwang) [inline pages/index.tsx:94]: i think the second parameter is optional, should we just omit it? or does having an empty object serve a purpose?

File: pages/index.tsx [hunk 24]
```diff
@@ -19,21 +34,136 @@ export default function Home(): ReactElement {
...
+  // Takes certificate that getCertificate function returns as argument,
+  // makes API call, returns all API data.
+  async function makeRequest(certificate: Pkcs12ReadResult) {
+    const headers = {
+      Accept: 'application/json',
+    }
+
+    /* https://nodejs.org/api/tls.html#tls_tls_createsecurecontext_options
+
+    // TODO: Store certs in memory, instead of loading every time.
```
Comment (kalvinwang) [inline pages/index.tsx:107]: This is already in a ticket so I think we can remove the todo here?

File: pages/index.tsx [hunk 25]
```diff
@@ -19,21 +34,136 @@ export default function Home(): ReactElement {
...
+    // TODO: rejectUnauthorized - set to true or implement a `checkServerIdentity`
```
Comment (kalvinwang) [inline pages/index.tsx:108]: Is this already in a ticket? If not it probably should be

File: pages/index.tsx [hunk 26]
```diff
@@ -19,21 +34,136 @@ export default function Home(): ReactElement {
...
+      key: certificate.key,
+      rejectUnauthorized: false,
+      keepAlive: false,
+    }
+
+    // Instantiate agent to use with TLS Certificate.
+    // Reference: https://sebtrif.xyz/blog/2019-10-03-client-side-ssl-in-node-js-with-fetch/
+    const sslConfiguredAgent: https.Agent = new https.Agent(options)
+
+    // TODO: if no uniqueNumber, redirect.
```
Comment (kalvinwang) [inline pages/index.tsx:124]: Let's ticket this? If there's no unique number and this is being accessed via the user-facing claim status url, this should be a Sev 3 alert.

**Review by kalvinwang — APPROVED** [review]
lgtm! Just some minor edits and questions about making tickets from todos. Also maybe add one more ticket for cleaning up types in index.tsx?

File: pages/index.tsx (hunk 24 above)
Comment (rocketnova) [reply pages/index.tsx:107]: Sounds good. I'll remove this TODO

File: pages/index.tsx (hunk 25 above)
Comment (rocketnova) [reply pages/index.tsx:108]: Yup. This is #182. I'll remove this TODO as well.

File: pages/index.tsx (hunk 26 above)
Comment (rocketnova) [reply pages/index.tsx:124]: Good call. I've created #217, so I'll remove this TODO

File: pages/index.tsx (hunk 23 above)
Comment (rocketnova) [reply pages/index.tsx:94]: Good call. I'll omit it.

File: pages/index.tsx [hunk 27]
```diff
@@ -1,16 +1,31 @@
...
+import pem, { Pkcs12ReadResult } from 'pem'
+import fetch, { Response } from 'node-fetch'
 
 import { Header } from '../components/Header'
 import { Main } from '../components/Main'
 import { Footer } from '../components/Footer'
 
-export default function Home(): ReactElement {
+export interface Claim {
+  ClaimType: string | 'not working'
```
Comment (kalvinwang) [inline pages/index.tsx:21]: should this be ?? '' for now and error handled later?
